lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -456,14 +477,29 @@ void FIRRTLModuleLowering::runOnOperation() {
         });
   }
 
+  // Now update all the testbench modules' paths.  A module goes in the TB
+  // directory if it isn't a child of the DUT and a DUT is marked.
+  if (tbdir) {
+    if (auto dut = state.getDut()) {
+      for (auto mod : state.oldToNewModuleMap) {
+        if (!state.getInstanceGraph()->isAncestor(mod.first, dut)) {
+          auto outputFile = hw::OutputFileAttr::getAsDirectory(
+              circuit.getContext(), tbdir.getValue(), false, true);
+          mod.second->setAttr("output_file", outputFile);
+        }
+      }
+    }
```

Comment:
It would be cool if there to do a depth first search from the top-level-module and  prune at the the DUT to find just the TB modules.  There might be a [way](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/DepthFirstIterator.h#L185)...

Reply:
Likewise this compiles, not sure what you will think of it though.  I did notice you only have to create `outputFile` once and can move it out of the loop.
 
```c++
  if (tbdir) {
    if (auto dut = state.getDut()) {
      auto outputFile = hw::OutputFileAttr::getAsDirectory(
          circuit.getContext(), tbdir.getValue(), false, true);
      auto i = llvm::df_begin(state.getInstanceGraph());
      auto e = llvm::df_end(state.getInstanceGraph());
      while (i != e) {
        auto oldModule = i->getModule();
        if (oldModule == dut) {
          // Don't increment the iterator when you skip children.
          i.skipChildren();
          continue;
        }
        auto newModule = state.oldToNewModuleMap[oldModule];
        newModule->setAttr("output_file", outputFile);
        ++i;
      }
    }
  ```

Reply:
I'm not assuming a tree.  It also isn't clear to me what to do if a module is both under the dut and also the TB.  I decided that the correct thing to do is leave them in the DUT dir, which I think this is doing.